core/src/main/scala/com/mesosphere/package.scala

```diff
@@ -0,0 +1,161 @@
+package com
+
+import com.typesafe.scalalogging.CanLog
+import com.typesafe.scalalogging.LoggerTakingImplicit
+import org.slf4j.MDC
+import scala.language.implicitConversions
+
+package object mesosphere {
+
+  /**
+   * Abstractions in this class were designed with below interests:
+   * 1. Use the slf4j abstraction without any backend (such as logback etc.,). Consumers of this library should be
+   *    free to chose whatever backend they wish to use. `MDC` is a well-known concept in java world and many logging
+   *    backends will be able to do a structured logging that includes MDC.
+   * 2. Make it easy to compose loggers with context. E.g.: [[Traversable]] of tuples of type ([[String]], [[Any]]).
+   * 3. User should be able to :
+   *     - create a logger instance `A` with few context parameters associated with it.
+   *     - Use the above logger for making log statements with the context built.
+   *     - Pass the instance `A` to another method or class which has no information about context built so far
+   *       but can chose to built context further on top of the current context OR start fresh.
+   *
+   * Usability primer:
+   *
+   * The types declared here such as [[KvArgs]] and [[LoggerAndContext]] can be used through out the project.
+   * Following snippet shows an example of how the [[LoggerTakingImplicit]] can be used with the
+   * abstractions defined here :
+   *
+   * {{{
+   *    // import all the implicit defined here in to scope.
+   *    import com.mesosphere._
+   *
+   *    // Create an instance of LoggerTakingImplicit.
+   *    // This needs an evidence of CanLog[KvArgs] to compile which was brought in to scope by above import.
+   *    val log = Logger.takingImplicit[KvArgs](getClass)
+   *
+   *    // Make a log statement with no context.
+   *    // This needs an implicit context but the above import bought an empty context in to scope.
+   *    log.info("hello world!")
+   *
+   *    // Make a log statement with some context.
+   *    log.info("hello world!")(("key1", "value1"))
+   *
+   *    // Make a log statement with more context.
+   *    log.info("hello world!")(("key1", "value1").and("key2", "value2").and("key3", "value3"))
+   *
+   *    // Now compose the logger with static meta data.
+   *    // The withContext method here is defined on a value class of LoggerTakingImplicit.
+   *    val composedLogger : LoggerAndContext = log.withContext("constantKey", "constantValue")
+   *
+   *    // Note that the type of composedLogger is just a simple tuple containing the logger as well as context.
+   *    // The state of logger object itself is not mutated in any manner.
+   *    // The logger is not knowledgeable about the context in any manner. The tuple is the ONLY thing that
+   *    // is used correlate between a logger and the context associated with it.
```

Comment:
What is the motivation of this. It feels we almost get the downside of both worlds. No real wrapper and no real context attached.

Why not 

```scala
class LoggerAndContext(logger: LoggerTakingImplicit[KvArgs], context: KvArgs) {

  def info() = ??? // how is the context passed?
}
```

Which would make the context passing explicit. Right now I still don't understand how

```scala
log.info("hello world!")(("key1", "value1"))
```

works.

Reply:
@jeschkies I think the data structure we need is a list of tuples